using_restricted_array.rb

```diff
@@ -6,44 +6,115 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
+
 def length(array)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] == nil
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] == nil
+    print "#{array[index]} "
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] > length
+    if array[index] == value_to_find
+      return true
+    else
+      return false
+      index =+ 1
+    end
+  end
 end
 
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
+
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  a = 0
+  while a < length
+    b = 0
+    while b < length-a-1
+      if array[b] > array[b+1] # swap
```

Comment:
Why do you need to swap to find the largest element? Please review the slide deck again and slack me if you have questions.

Reply:
Aha! I see - you are doing a bubble sort to sort the array. This is more expensive. Bubble sort is an O(n^2) algorithm. So the time complexity of your find_largest is currently O(n^2). How can you make your find_largest algorithm on an unsorted array be an O(n) algorithm? Review slides to revise how we did this in class.